### PR TITLE
MM-53646: First version of the release process

### PR DESCRIPTION
**Pull request by agarciamontoro**
#### Summary
This PR proposes a new release process, with a fixed timeline and owners, to ensure released versions are well tested and stable. It is simple by design, we can always iterate on it when needed.

Adding everyone in the team that may be part of it as stated in the doc as reviewers.

Everything in this document is up to discussion, there's nothing set in stone, so feel free to disagree with some or even all of it :)

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53646

### CODE REVIEW

File: docs/release.md
Comment (agarciamontoro) [inline docs/release.md:78]: This part is just moved from `developing.md`, it now makes more sense to have it here.

File: docs/release.md
Comment (streamer45) [inline docs/release.md:14]: Super unrelated comment but this is great. Didn't know footnotes were supported.

File: docs/release.md
Comment (streamer45) [inline docs/release.md:18]: Historically this tool has seen periods of high and low activity hence why we never formalized on a fixed cadence.
I am thinking there's a chance we could end up skipping releases in the future due to nothing changing in the month. Not a huge deal, just something to be aware of, unless we want to bump version for the sake of it but not sure it's needed, unless of course we decide to start mapping versions with the product (e.g. `mattermost-load-test-ng` 9.0, 9.1 etc).

**Review by streamer45 — COMMENTED** [review]
Thanks, left a few comments.

A meta thought would be whether we have considered using a playbook to guide the various stakeholders through this recurring process.

**Review by agnivade — APPROVED** [review]

File: docs/release.md
Comment (agnivade) [reply docs/release.md:49]: Sort by age!

File: docs/release.md
Comment (agarciamontoro) [reply docs/release.md:18]: > I am thinking there's a chance we could end up skipping releases in the future due to nothing changing in the month. 

Indeed, I'll clarify that skipping releases should be expected if no changes happened.

> unless we want to bump version for the sake of it but not sure it's needed, unless of course we decide to start mapping versions with the product

I don't think that's needed, no.

File: docs/release.md
Comment (agarciamontoro) [reply docs/release.md:24]: I think we should be strict, yes. Whatever we have at that point, we cut the branch. If there are no changes, we skip the release. So it will be at this point when we decide whether we'll need a release or not.

File: docs/release.md
Comment (agarciamontoro) [reply docs/release.md:34]: The real performance comparison will still happen, and they should block the Mattermost release if needed. What I meant here is that if we don't happen to have fixed the release *of the load-test tool* at this point (which we have tested with a dummy performance comparison), we will simply use the latest stable version for the real performance comparison, instead of trying to rush the new one in.

**Comment by agarciamontoro** [comment]
> A meta thought would be whether we have considered using a playbook to guide the various stakeholders through this recurring process.

Sure! I'll create one and link it here.

**Comment by agarciamontoro** [comment]
Done: https://community.mattermost.com/playbooks/playbooks/95pa6a8fcp8p8xsn4frqejcide

**Review by amyblais — APPROVED** [review]

File: docs/release.md
Comment (streamer45) [reply docs/release.md:36]: Yes, but in case of new coverage the two things are sort of coupled. We cannot really use a previous version of the tool to test performance right?

Comment (agarciamontoro) [reply docs/release.md:36]: Indeed. This is the problem we always face: we need to have coverage of new features, and we need to have a stable tested version of the tool to run the performance comparisons. Feature complete for Mattermost is, according [to the handbook](https://handbook.mattermost.com/operations/research-and-development/product/release-process/feature-release), "approximately 1 month prior to the release day", so T-30. If that rule is followed, and if devs add the corresponding coverage to the load-test tool *at the same time* as the feature is being developed, then we should be good.
That's a lot of *if*s, though. We could try to shorten the testing/debugging/fixing period defined here, maybe leave it in one week, so we have an additional week for finishing the coverage after feature complete in Mattermost.
Thoughts?

**Review by isacikgoz — APPROVED** [review]

File: docs/release.md
Comment (streamer45) [reply docs/release.md:36]: I think we shouldn't be able to merge any new complex functionality in monorepo until performance tests have been conducted on the PR which means coverage becomes a requirement. I think that should be the standard process, not the opposite (merging and hoping coverage will get in time). It will inevitably mean that some features will get delayed due to this and that's in the nature of things. What we cannot do going forward (in my opinion) is shipping untested software.

Comment (agarciamontoro) [reply docs/release.md:36]: I agree. This is something that every dev should be aware of. And while we've been seeing more and more coverage added to the tool by other teams, I'm sure we're still missing PRs that may benefit from this.
But let's assume, so we don't increase the scope of this discussion greatly, that this is already happening. For the release process being discussing here, that would mean that we *need* to release a new version (if there are indeed changes) and may end up delaying the product release if we have some unfixed bugs by T-15. I think the step in the release would be something like "communicate to the release manager that the release process will be delayed". I guess at this point we would enter fire-extinguishing mode and try to get the release as soon as possible so that we can run the comparison.
An alternative of course is to be more strict in our own PRs, and thoroughly test all of them (accepting the delay that will cause).
Thoughts?

Comment (streamer45) [reply docs/release.md:36]: Yes, I think that's reasonable. Also if there's a specific regression in the tool itself unrelated to newly added coverage we could always consider reverting to unblock the release/testing process. I think it may depend on case by case basis but overall I am supportive of your proposal.

Comment (agarciamontoro) [reply docs/release.md:36]: > if there's a specific regression in the tool itself unrelated to newly added coverage we could always consider reverting to unblock the release/testing process

Yeah, that was part of the "fire-extinguishing mode" in my mind :) Ok, I'll reword it tomorrow and ask for a review again. Thanks for the discussion!

File: docs/release.md
Comment (streamer45) [inline docs/release.md:57]: Maybe we could specify it's the binary release or bundle we are referring to here. Mostly because we are using the word *release* from different levels.

File: docs/release.md
Comment (streamer45) [inline docs/release.md:3]: Should we specify which team to ping or do you think it's sufficiently covered by the owners below?

File: docs/release.md
Comment (streamer45) [inline docs/release.md:10]: Just some random thought at this point but have we ever considered doing an extra comparison for ESRs to be compared with the previous ESRs?

File: docs/release.md
Comment (streamer45) [inline docs/release.md:12]: I assume this is implicitly understood by everyone here but just want to highlight that there's always the case RC-1 won't be enough if we found a performance regression that either needs fixing or reverting.

**Review by streamer45 — COMMENTED** [review]
Great stuff, left a couple more comments.

One extra piece of feedback that comes out of some of the discussion above is that maybe there's also an opportunity to improve our official release process in the handbook if we can find an agreement between the various stakeholders that a successful load-test comparison step becomes a strict requirement to ship any new MM version. I think it could be a great step towards more awareness on the importance of performance.

File: docs/release.md
Comment (agarciamontoro) [reply docs/release.md:57]: Do you mean the ltagent/ltapi vs ltctl? At some point we should release ltctl as well I think, otherwise it'll be confusing.

File: docs/release.md
Comment (agarciamontoro) [reply docs/release.md:3]: Ah, yes, I'll specify it!

File: docs/release.md
Comment (agarciamontoro) [reply docs/release.md:10]: That's something that's been on my mind for a while, but we've never planned anything related. We probably should at some point.

Comment (amyblais) [reply docs/release.md:10]: I agree with the idea of adding loadtesting to compare ESRs.

File: docs/release.md
Comment (streamer45) [reply docs/release.md:57]: I mean the Github release bundle, whatever that may contain.

File: docs/release.md
Comment (streamer45) [reply docs/release.md:12]: > Do you mean that the finally released version may be RC-2 or higher?

Yeah, and it may imply having to do a second comparison as well.

Comment (agarciamontoro) [reply docs/release.md:12]: I was gonna clarify this in the document, but on second thought, that would be part of the performance comparison process, not part of the release process. Once the load-test version is released, we will use that one to run the performance comparison as many times as needed on RC-1and following. If we ever need to revert a feature from the load-test tool after RC-1, we can plan that ad-hoc.

File: docs/release.md
Comment (agarciamontoro) [reply docs/release.md]: https://github.com/mattermost/mattermost-load-test-ng/pull/640/commits/666a4a3d75a1a1b55a6345642adbe4d28892545c

File: docs/release.md
Comment (agarciamontoro) [reply docs/release.md]: https://github.com/mattermost/mattermost-load-test-ng/pull/640/commits/9d18f1b7f1f03a463e4e599dad9c8d23229b98d4

**Comment by agarciamontoro** [comment]
> A note on Claudio's comment for: "maybe there's also an opportunity to improve our official release process in the handbook if we can find an agreement between the various stakeholders that a successful load-test comparison step becomes a strict requirement to ship any new MM version."
> I agree with this as well and I'm happy to update the handbook if there is agreement from other dev teams as well.

I'll bring this up in the dev meeting today and see if there's anything against it.

**Comment by agarciamontoro** [comment]
I'm merging this PR with the hope of trying the release process synchronized with the release of v9.2, which starts tomorrow. The dates will be as follow:

- T-24 (October 13): Decide if a new release is worth happening.
- T-23 to T-12 (October 16 to October 31): Testing period
- T-11 (November 1): Release new version
- T-10 to T-8 (November 2 to November 6): Performance comparison for v9.2

**Pull request merged**
